src/v/cloud_storage/remote.cc

```diff
@@ -782,6 +818,73 @@ ss::future<upload_result> remote::delete_objects(
     co_return *result;
 }
 
+ss::future<upload_result> remote::delete_objects_sequentially(
+  const cloud_storage_clients::bucket_name& bucket,
+  std::vector<cloud_storage_clients::object_key> keys,
+  retry_chain_node& parent) {
+    vlog(
+      cst_log.debug,
+      "Backend {} does not support batch delete, falling back to "
+      "sequential deletes",
+      _cloud_storage_backend);
+
+    std::vector<key_and_node> key_nodes;
+    key_nodes.reserve(keys.size());
+
+    std::transform(
+      std::make_move_iterator(keys.begin()),
+      std::make_move_iterator(keys.end()),
+      std::back_inserter(key_nodes),
+      [&parent](auto&& key) {
+          return key_and_node{
+            .key = std::forward<cloud_storage_clients::object_key>(key),
+            .node = std::make_unique<retry_chain_node>(&parent)};
+      });
+
+    auto results = co_await ss::do_with(
+      bucket,
+      std::move(key_nodes),
+      std::vector<upload_result>{},
+      [this](auto& bucket, auto& key_nodes, auto& results) {
+          results.reserve(key_nodes.size());
+          return ss::max_concurrent_for_each(
```

Comment:
why are we using `do_with` in a coroutine?

Reply:
I started out with a `do_with`, because `max_concurrent_for_each` needs the key node vector to be alive for all calls, and if we use a coroutine then the lambda passed as the final argument for `max_concurrent_for_each` could cause lifetime problems IIUC.

Then I added the `co_await` at the result of `ss:do_with` for convenience, although we can also chain the result reduction after `do_with`. 

Is the `co_await` with `do_with` going to be a problem in this context? I assumed since there is no lambda used with the `co_await` and we are using `do_with` to make objects live long enough it should be fine? 

If not I can change it to add a continuation which returns an actual result and remove the coroutine.

Reply:
> Is the co_await with do_with going to be a problem in this context? 

it won't cause a problem, i just couldn't figure out why it would be needed. it looked like you could call co_await on max_concurrent_for_each directly with `bucket`, `key_nodes`, and `results` all stored on the coroutine frame. no need to change it, just curious.

Reply:
`max_concurrent_for_each` will need to take a lambda, which would also be a coroutine. To avoid potential lifetime issues with [lambda+coroutine](https://github.com/scylladb/seastar/blob/master/doc/lambda-coroutine-fiasco.md) I used explicit `do_with`. 

But now looking at `max_concurrent_for_each`  source https://github.com/scylladb/seastar/blob/master/include/seastar/core/loop.hh#L692 it already uses `do_with` to store the state (the iterators, the upper bound count and the function) on the heap. 

So perhaps using `co_await` is fine here.

Reply:
i think so. the outer co_await / coroutine can keep data alive even if the inner lambda running with max_concurrent_for_each is not a coroutine but reference the extended lifetime data. anyway, no need to change